### ISSUE TITLE
@LoadBalanced WebClient doesnt work with WebTestClient

### ISSUE THREAD

**Issue by dushshantha** [spring-cloud/spring-cloud-commons#570]
I have a few tests annotated as @RestClientTest that uses a @LoadBalanced WebClient.Builder to call a service. I have a MockWebServer running to mock the service on the correct port. I use Ribbon for load balancing. I get the following exception when the tests run:

```
eactor.core.Exceptions$ReactiveException: java.net.UnknownHostException: orchestration

at reactor.core.Exceptions.propagate(Exceptions.java:326)
at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:91)
at reactor.core.publisher.Mono.block(Mono.java:1494) ...
```
My Bean definition:

```
@LoadBalanced
public WebClient.Builder asyncWebClientBuilder() {
    TcpClient tcpClient = TcpClient.create()
            .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, connectTimeout) // Connection Timeout
            .doOnConnected(connection ->
                    connection.addHandlerLast(new ReadTimeoutHandler(readTimeout))); // Read Timeout)
    return WebClient.builder()
            .clientConnector(new ReactorClientHttpConnector(HttpClient.from(tcpClient)));
```
The Client Code:

```
return client.build()
            .method(method)
            .uri(initUriTemplateHandler()
                    .expand(url, uriVariables))
            .body(BodyInserters.fromObject(requestEntity))
            .exchange()
            .block();
```
Strange thing is, the same exact client code works perfectly when I run it in the real world situation against the real service. I also have a few tests that are @SpringBootTest annotated and use the same exact client underneath which load balanced and works perfectly. It's not load balanced only for the case of @RestClientTest case. I am using Spring Boot 2.1.3

**Comment by OlgaMaciaszek** [comment]
Hi @dushshantha I guess you mean [WebTestClient](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/test/web/reactive/server/WebTestClient.html) and not @RestClientTest?

Please provide more details regarding how you set up the `client` instance in your tests and concerning your application context and Ribbon setup.

**Comment by dushshantha** [comment]
Hi @OlgaMaciaszek,
Thanks for the response. Actually, I meant @RestClientTest annotation for TestNG. Let me give you the background of my issue.

In my Spring Boot 2.1.3 application, I have a requirement to call a few external REST APIs. To handle these, I created few Client classes that uses WebClient in them. The WebClient bean is defined as @LoadBalanced as my external APIs are load Balanced.

Now, to test my Client code, I have a few TestNG test cases. These test classes ate annotated as [@RestClientTest](https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/test/autoconfigure/web/client/RestClientTest.html) 

In my test case, All I do is calling the method in my Client class which will call the API using WebClient for me. My test looks like this.

```
public void stepOkTest() throws IOException {
    // Here I start up a MockWebServer to mock the API.

    mockServerCalls.addUaaCall();
  
    // Calling the client which will call WebClient.build().eschange() as I mentioned in the original description.  
    boolean result = client.execute();

    Assert.assertTrue(result);
}
```
 Inside the execute() method, I have a code similar to this:

```
return client.build()
            .method(method)
            .uri(initUriTemplateHandler()
                    .expand(url, uriVariables))
            .body(BodyInserters.fromObject(requestEntity))
            .exchange()
            .block();
```

where the client is a WebClient.Builder

The URL i am passing in is a LoadBalancer url, for example http://loadbalancedname/test instead of a areal URL. expectation is for the LoadBalancer Ribbon to resolve loadbalancedname to the correct URL. Now this exact code works just fine if I use RestTemplate as the client instead of WebClient. As soon as I use WebClient to call the API, LoadBa;ancer stops resolving the correct url. Hope this makes sense

**Comment by OlgaMaciaszek** [comment]
Thanks for the info @dushshantha . 
Please provide a [minimal, complete, verifiable example](https://stackoverflow.com/help/mcve) that reproduces the issue.

**Comment by dushshantha** [comment]
Hi,
I found a solution for this. My issue was that when I was using @RestClientTest annotation for my test classes, It did not load the Spring Boot Application context was not loaded, which is what executes Ribbon to load Balance. My application was in Spring Boot 1.5 and I just upgraded it to 2.1.3 along with using WebClient instead of RestTemplate. I changed my tests to use @SpringBootTest and Ribbon started working for the tests.
